dbms/src/Interpreters/TraceCollector.cpp

```diff
@@ -0,0 +1,58 @@
+#include "TraceCollector.h"
+
+#include <common/Sleep.h>
+#include <common/StackTrace.h>
+#include <common/logger_useful.h>
+#include <IO/ReadHelpers.h>
+#include <IO/ReadBufferFromFileDescriptor.h>
+#include <Common/Exception.h>
+#include <Common/QueryProfiler.h>
+#include <Interpreters/TraceLog.h>
+
+namespace DB
+{
+
+    TraceCollector::TraceCollector(std::shared_ptr<TraceLog> trace_log)
+        : log(&Logger::get("TraceCollector"))
+        , trace_log(trace_log)
+    {
+    }
+
+    void TraceCollector::run()
+    {
+        DB::ReadBufferFromFileDescriptor in(trace_pipe.fds_rw[0]);
+
+        while (true)
+        {
+            SleepForMicroseconds(1);
```

Comment:
period should be configurable

Reply:
Why do we need to sleep at all?
We will block on reading from pipe. It is the synchronization primitive.

Reply:
Sure